sheets/batch_get.go

```diff
@@ -0,0 +1,162 @@
+	sheetService, err := sheets.NewService(ctx, option.WithHTTPClient(args.OAuthConfig.Client(ctx, args.OAuthToken)))
+	if err != nil {
+		return nil, fmt.Errorf("error creating sheets service client: %w", err)
+	}
+	return &BatchReader{
+		spreadsheetID:        args.SpreadsheetID,
+		sheetID:              args.SheetID,
+		pollingPeriod:        args.PollingPeriod,
+		sheetSvc:             sheetService,
+		dateTimeRenderOption: args.DateTimeRenderOption,
+		valueRenderOption:    args.ValueRenderOption,
+	}, nil
+}
+
+// GetSheetRecords returns the list of records up to a maximum of 1000 rows(default limit)
+// added after the row offset of last successfully read record
+func (b *BatchReader) GetSheetRecords(ctx context.Context, offset int64) ([]sdk.Record, error) {
+	if b.nextRun.After(time.Now()) {
+		return nil, nil
+	}
+
+	res, err := b.sheetSvc.Spreadsheets.Values.BatchGetByDataFilter(b.spreadsheetID, b.getDataFilter(offset)).Context(ctx).Do()
+	if err != nil {
+		if googleapi.IsNotModified(err) {
+			return nil, nil
+		}
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusTooManyRequests {
+			b.retryCount++
+			duration := time.Duration(b.retryCount * int64(b.pollingPeriod)) // exponential back off
+			b.nextRun = time.Now().Add(duration)
+			sdk.Logger(ctx).Error().
+				Int64("retry_count", b.retryCount).
+				Float64("wait_duration", duration.Seconds()).
+				Msg("exponential back off, rate limit exceeded")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("error getting sheet(gid:%v) values, %w", b.sheetID, err)
+	}
+
+	b.retryCount = 0
+	return b.valueRangesToRecords(res.ValueRanges, offset)
+}
+
+func (b *BatchReader) getDataFilter(offset int64) *sheets.BatchGetValuesByDataFilterRequest {
+	dataFilters := make([]*sheets.DataFilter, 0)
+	dataFilters = append(dataFilters, &sheets.DataFilter{
+		GridRange: &sheets.GridRange{
+			SheetId:       b.sheetID,
+			StartRowIndex: offset,
+		},
+	})
+	return &sheets.BatchGetValuesByDataFilterRequest{
+		DataFilters:          dataFilters,
+		DateTimeRenderOption: b.dateTimeRenderOption,
+		MajorDimension:       majorDimension,
+		ValueRenderOption:    b.valueRenderOption,
+	}
+}
+
+func (b *BatchReader) valueRangesToRecords(valueRanges []*sheets.MatchedValueRange, offset int64) ([]sdk.Record, error) {
+	records := make([]sdk.Record, 0)
+	// iterate over all the value ranges fetched from the Google sheet BatchGet API request
+	for i := range valueRanges {
+		values := valueRanges[i].ValueRange.Values
+		// Iterate over the Rows of the value range
+		// Data is of format: [][]interface{} => ([ [ROW1 => A1,B1,C1..], [ROW2 => A2, B2, C2,...],...])
+		for index, val := range values {
+			if len(val) == 0 {
+				continue
+			}
+			rawData, err := json.Marshal(val)
+			if err != nil {
+				return records, fmt.Errorf("error marshaling the map: %w", err)
+			}
+			rowOffset := offset + int64(index) + 1
```

Comment:
Does this still work, if we for example delete all the rows in a sheet? E.g. we have a sheet with 10 rows, source pulls them all, then we delete all of them, and add 2 more. Will those two last ones be picked up?

Reply:
Nope, we are not able to fetch the updated rows from google sheets, so only new rows will be tracked.
In the above case, new records will be tracked only from row 11

Reply:
We have mentioned it in known limitations:
https://github.com/gopherslab/conduit-connector-google-sheets/pull/3/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R115